bee-protocol/src/worker/transaction/hasher.rs

```diff
@@ -0,0 +1,188 @@
+
+use std::pin::Pin;
+
+// If a batch has less than this number of transactions, the regular CurlP hasher is used instead
+// of the batched one.
+const BATCH_SIZE_THRESHOLD: usize = 3;
+
+type Receiver = ShutdownStream<mpsc::UnboundedReceiver<HasherWorkerEvent>>;
+
+pub(crate) struct HasherWorkerEvent {
+    pub(crate) from: EndpointId,
+    pub(crate) transaction: TransactionMessage,
+}
+
+#[pin_project(project = HasherWorkerProj)]
+pub(crate) struct HasherWorker {
+    processor_worker: mpsc::UnboundedSender<ProcessorWorkerEvent>,
+    #[pin]
+    receiver: Receiver,
+    cache: HashCache,
+    hasher: BatchHasher,
+    events: Vec<HasherWorkerEvent>,
+}
+
+impl HasherWorker {
+    pub(crate) fn new(
+        processor_worker: mpsc::UnboundedSender<ProcessorWorkerEvent>,
+        cache_size: usize,
+        receiver: Receiver,
+    ) -> Self {
+        assert!(BATCH_SIZE_THRESHOLD <= BATCH_SIZE);
+        Self {
+            processor_worker,
+            receiver,
+            cache: HashCache::new(cache_size),
+            hasher: BatchHasher::new(8019, CurlPRounds::Rounds81),
+            events: Vec::with_capacity(BATCH_SIZE),
+        }
+    }
+
+    async fn trigger_hashing(&mut self, batch_size: usize) {
+        if batch_size < BATCH_SIZE_THRESHOLD {
+            let hashes = self.hasher.hash_unbatched();
+            Self::send_hashes(hashes, &mut self.events, &mut self.processor_worker).await
+        } else {
+            let hashes = self.hasher.hash_batched();
+            Self::send_hashes(hashes, &mut self.events, &mut self.processor_worker).await
+        }
+        // FIXME: we could store the fraction of times we use the batched hasher
+    }
+
+    async fn send_hashes(
+        hashes: impl Iterator<Item = TritBuf>,
+        events: &mut Vec<HasherWorkerEvent>,
+        processor_worker: &mut mpsc::UnboundedSender<ProcessorWorkerEvent>,
+    ) {
+        for (HasherWorkerEvent { from, transaction }, hash) in events.drain(..).zip(hashes) {
+            if let Err(e) = processor_worker.unbounded_send(ProcessorWorkerEvent {
+                hash: Hash::from_inner_unchecked(hash),
+                from,
+                transaction,
+            }) {
+                warn!("Sending event to the processor worker failed: {}.", e);
+            }
+        }
+    }
+
+    pub async fn run(mut self) -> Result<(), WorkerError> {
+        info!("Running.");
+
+        while let Some(batch_size) = self.next().await {
+            self.trigger_hashing(batch_size).await
+        }
+
+        info!("Stopped.");
+
+        Ok(())
+    }
+}
+
+impl Stream for HasherWorker {
+    type Item = usize;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        // We need to do this because `stream` needs to be pinned to be polled.
+        let HasherWorkerProj {
+            mut receiver,
+            hasher,
+            events,
+            cache,
+            ..
+        } = self.project();
```

Comment:
How does that work ? Does it make a copy of everything ?

Reply:
those are just references to the original fields.

Reply:
Thanks